libraries/core_libs/consensus/src/transaction/transaction_manager.cpp

```diff
@@ -312,19 +312,23 @@ SharedTransactions TransactionManager::getAllPoolTrxs() {
  */
 void TransactionManager::updateFinalizedTransactionsStatus(PeriodData const &period_data) {
   // !!! There is no lock because it is called under std::unique_lock trx_lock(trx_mgr_->getTransactionsMutex());
-  last_finalized_block_transactions_.clear();
-  for (auto const &trx : period_data.transactions) {
-    last_finalized_block_transactions_[trx->getHash()] = trx;
-    if (!nonfinalized_transactions_in_dag_.erase(trx->getHash())) {
-      trx_count_++;
-    } else {
-      LOG(log_dg_) << "Transaction " << trx->getHash() << " removed from nonfinalized transactions";
+  if (period_data.transactions.size() > 0) {
+    if (period_data.transactions.size() + recently_finalized_transactions_.size() > kRecentlyFinalizedTransactionsMax) {
```

Comment:
Won't ExpirationCacheMap work better here?

Reply:
ExpirationCacheMap is thread safe and since all of this is already under transactions mutex we would be double locking. Probably some kind of more specialized cache would be better here since the current solution at some poiint completely clears the cache but this is just something I wanted to test on devnet to confirm that this is actually the bottleneck for high TPS and then once I confirm it I would try to implement better caching